### PR TITLE
CEDS-3362 XML to Exports Data Model mapping - Locations

### PR DESCRIPTION
**Pull request by MaciejRewera**

### CODE REVIEW

File: app/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/DestinationCountryParser.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import uk.gov.hmrc.exports.services.reversemapping.declaration.XmlTags._
+
+import javax.inject.Singleton
+import scala.xml.NodeSeq
+
+@Singleton
+class DestinationCountryParser extends DeclarationXmlParser[Option[Country]] {
+
+  override def parse(inputXml: NodeSeq)(implicit context: MappingContext): XmlParserResult[Option[Country]] = Right(
+    StringOption((inputXml \ Declaration \ GoodsShipment \ Destination \ CountryCode).text)
```
Comment (glidester) [inline app/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/DestinationCountryParser.scala:33]: Should this not use Lucio's new `toStringOption`?

File: app/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/OriginationCountryParser.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import scala.xml.NodeSeq
+
+@Singleton
+class OriginationCountryParser extends DeclarationXmlParser[Option[Country]] {
+
+  override def parse(inputXml: NodeSeq)(implicit context: MappingContext): XmlParserResult[Option[Country]] = Right(
+    StringOption((inputXml \ Declaration \ GoodsShipment \ ExportCountry \ ID).text)
+      .map(exportCountryId => Country(Some(exportCountryId)))
+  )
+}
```
Comment (glidester) [inline app/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/OriginationCountryParser.scala:36]: Nit: maybe you could combine this and the DestinationCountryParser into a single common CountryParser class that is passed in a filtered NodeSeq by the calling LocationsParser (similar to how PartiesParser calls the common EntityDetailsParser class multiple times for different values)

File: test/unit/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/DestinationCountryParserSpec.scala [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+class DestinationCountryParserSpec extends UnitSpec with EitherValues {
+
+  private val parser = new DestinationCountryParser
+
+  private implicit val mappingContext = MappingContext(eori = ExportsTestData.eori)
+
+  "DestinationCountryParser on parse" should {
+
+    "return None" when {
+      "the 'GoodsShipment / Destination / CountryCode' element is NOT present" in {
```
Comment (glidester) [inline test/unit/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/DestinationCountryParserSpec.scala:36]: Should also test for when element is present but value is empty

File: test/unit/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/DestinationCountryParserSpec.scala [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+class DestinationCountryParserSpec extends UnitSpec with EitherValues {
+
+  private val parser = new DestinationCountryParser
+
+  private implicit val mappingContext = MappingContext(eori = ExportsTestData.eori)
+
+  "DestinationCountryParser on parse" should {
+
+    "return None" when {
```
Comment (glidester) [inline test/unit/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/DestinationCountryParserSpec.scala:35]: Should also test for when element is present but value is empty

File: test/unit/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/GoodsLocationParserSpec.scala [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+class GoodsLocationParserSpec extends UnitSpec with EitherValues {
+
+  private val parser = new GoodsLocationParser
+
+  private implicit val mappingContext = MappingContext(eori = ExportsTestData.eori)
+
+  "GoodsLocationParser on parse" should {
+
+    "return Right with None" when {
```
Comment (glidester) [inline test/unit/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/GoodsLocationParserSpec.scala:36]: Should also test for when elements are present but values are empty

**Review by glidester — COMMENTED** [review]
A couple of comments. Tests classes should also check for cases when elements are present but their values are empty should return None.

File: app/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/GoodsLocationParser.scala [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  override def parse(inputXml: NodeSeq)(implicit context: MappingContext): XmlParserResult[Option[GoodsLocation]] =
+    Right((inputXml \ Declaration \ GoodsShipment \ Consignment \ GoodsLocationTag).headOption.flatMap(parseGoodsLocation))
+
+  private def parseGoodsLocation(goodsLocationElement: NodeSeq): Option[GoodsLocation] = {
+    val country = (goodsLocationElement \ Address \ CountryCode).toStringOption
+    val typeOfLocation = (goodsLocationElement \ TypeCode).toStringOption
+    val qualifierOfIdentification = (goodsLocationElement \ Address \ TypeCode).toStringOption
+    val identificationOfLocation = (goodsLocationElement \ Name).toStringOption
+
+    if (Seq(country, typeOfLocation, qualifierOfIdentification).flatten.isEmpty)
```
Comment (udinobi) [inline app/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/GoodsLocationParser.scala:40]: Out of curiosity... What if `identificationOfLocation` is `nonEmpty`?

Comment (MaciejRewera) [reply app/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/GoodsLocationParser.scala:40]: Didn't think about it... I'll add a test case for that and I'll change it to return incomplete `GoodsLocation` object. This is more in line with what is already implemented.

**Review by glidester — APPROVED** [review]
:+1:

**Review by udinobi — APPROVED** [review]
